src/lib.rs

```diff
@@ -218,10 +212,166 @@ use srml_support::traits::Currency;
 use srml_support::{decl_event, decl_module, decl_storage, dispatch, StorageValue};
 use system::ensure_signed;
 
-pub trait MembershipRegistry<T: system::Trait> {
-    fn ensure_member(who: T::AccountId) -> Result<(), &'static str>;
+use rstd::collections::btree_map::BTreeMap;
+
+/// Constant values
+/// Later add to st
+///
+
+/// Represents a user in this forum.
+pub trait ForumUser<AccountId> {
+    
+    /// Identifier of user 
+    fn id(&self) -> &AccountId;
+
+    // In the future one could add things like 
+    // - updating post count of a user
+    // - updating status (e.g. hero, new, etc.)
+    //
+
+}
+
+/// Represents a regsitry of `ForumUser` instances.
+pub trait ForumUserRegistry<AccountId> {
+
+    type ForumUser: ForumUser<AccountId>;
+    
+    fn get_forum_user(&self, id:AccountId) -> Option<&Self::ForumUser>;
+
+}
+
+/// Convenient composite time stamp 
+struct BlockchainTimestamp<BlockNumber, Moment> {
+    block : BlockNumber,
+    time: Moment
+}
+
+//#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+/// Represents a moderation outcome applied to a post or a thread. 
+struct ModerationAction<BlockNumber, Moment, AccountId> {
+
+    /// When action occured.
+    occured_at: BlockchainTimestamp<BlockNumber, Moment>,
+
+    /// Account forum sudo which acted.
+    forum_sudo_account_id: AccountId,
+
+    /// Moderation rationale
+    rationale: Vec<u8>
+
 }
 
+/// Represents a revision of the text of a Post
+struct PostTextEdit<BlockNumber, Moment> {
```

Comment:
1. I think this idea should have been raised in the spec phase. It's not a complete proposal yet, as it would need things like an edit counter in the `Post`, in order for a light client to know that the full set of edits have been provided.

2. You are mixing two ideas here. (a) One is whether the most recent text should live as a field directly in `Post`, and all the edits live in the vector field. (b) The other is whether the edits vector should itself be inside `Post`, or in a separate map.

For (a) I think that could be a slight improvement yes, as it just makes it slightly less complicated and/or resource intensive to read the post text securely, depending on how (b) is solved.

For (b) I think this also is a good idea, but it really only starts to be important when/if there starts being a large number of edits per post.

**Conclusion: ** I think your suggestion is the right approach long term, but for now let's just stick to what we have, as we have passed the spec stage, and at this time this approach causes no problems, as the entire vector of edits is downloaded anyway, so its trivial to access. Also, there will not be many edits for a long time.